### ISSUE TITLE
Suggestion: use plugin template and only import the services needed

### ISSUE THREAD

**Issue by pimlie** [lupas/nuxt-fire#63]
Hey, thanks for this module. I was wondering whether you included the service if's on purpose in the plugin or not? The issue at the moment is that all the firebase services will still be included in your bundles. Ie if you dont use firestore you will still have a firestore chunk of 1MB.

If you would refactor the plugin as follows (so the ifs are on the template level, not on the plugin level) then the firebase services that are not enabled during build time wont be included in any bundle

```js
import firebase from 'firebase/app'
<% if (options.services.auth) { %>import 'firebase/auth'<% } %>
<% if (options.services.realtimeDb) { %>import 'firebase/database'<% } %>

export default async (ctx, inject) => {
  const firebaseConfig = JSON.parse(`<%= serialize(options.config) %>`)

  // Don't include when Firebase is already initialized
  if (!firebase.apps.length) {
    firebase.initializeApp(firebaseConfig)
  }

  <% if (options.services.auth) { %>
  const fireAuth = firebase.auth()
  const fireAuthObj = firebase.auth
  inject('fireAuth', fireAuth)
  inject('fireAuthObj', fireAuthObj)
  <% } %>

  <% if (options.services.realtimeDb) { %>
  const fireDb = firebase.database()
  const fireDbObj = firebase.database
  inject('fireDb', fireDb)
  inject('fireDbObj', fireDbObj)
  <% } %>
...
```

**Comment by lupas** [comment]
Hey @pimlie 

That's exactly what we did before v2.2.0, see [here](https://github.com/lupas/nuxt-fire/commit/fac31ce9c18bd2caa7a67467578aed3b79ad6b94).

With [version 2.2.0](https://github.com/lupas/nuxt-fire/releases/tag/v2.2.0) we starting using dynamic imports that help with splitting the chunks as suggested by the Firebase team.

As you can see in the commit linked above we first still included the templating conditionals (`<% if %>`) but then later dropped them because we found after testing and analyzing that conditional imports were somehow not bundled anyway. (seems we were wrong)

But thinking back now that doesn't make much sense. I guess we did something wrong testing this, because looking at it now I can see even disabled services being added to the bundle.

Long story short:
You're right, that's exactly how this plugin should do it. However I still prefer to do the imports asynchronously to leverage that benefit of chunks being split.

I created a branch where I re-implemented this, could you have a look and let me know if that is how you would see it fit to be implemented?

https://github.com/lupas/nuxt-fire/tree/reusePluginTemplating

(If you like you can create a pull-request with that change so it's gonna count as your contribution.)

Thanks a lot for the input! :)

**Comment by pimlie** [comment]
Ah right, interesting. I guess it make sense to have dynamic imports for those 1M chunks ;)

Maybe you could make the dynamic imports optional / more configurable? Also for dynamic imports it could be interesting to support webpack magic comments: https://webpack.js.org/api/module-methods/#magic-comments

Say my project always requires fireAuth to be loaded, then I probably just want to include it in my main vendor chunk (no dynamic import) or at the very least make sure it has a _webpackPreload_ comment so the browser knows I need that chunk right away.

So probably what I'm suggesting is something like this:
```js
// default behaviour, dynamic import as currently
  services: {
    auth: true
  }

// configurable import behaviour
  services: {
    auth: {
      dynamic: true/false, // true is default, eg check with _'dynamic' in service_ to see if the user has configured a value
      preload: true/false, // only makes sense when dynamic: true
      chunkName: isDev ? 'firebase-auth' : '[id]', // only makes sense when dynamic: true
    }
  }
```

**Comment by lupas** [comment]
@pimlie 

That's some great input, thanks!
As you might have figured I'm not the webpack & chunk expert yet... but I'm getting there ;-)

I just updated the branch with what you suggested, and it works fine when I tested it, care to take a look?
https://github.com/lupas/nuxt-fire/tree/reusePluginTemplating

Two questions:
1) Wouldn't it make sense that nuxt-fire just names the chunks for what they are (e.g. "firebase-auth"), or is there some benefit that the user can named them to whatever?

2) Do you think it makes sense to also implement other webpack options such as `webpackPrefetch` or `webpackMode`while we're at it?

**Comment by pimlie** [comment]
Hey, this is mostly personal preference but when having 3+ if's that mostly do the same I often abstract them away in to some method as it improves code readability and maintainability. Eg something like the following (untested):
```js
<%
const serviceMapping = {
  realtimeDb: 'database',
  ...
}

function writeStaticImports() {
  return Object.keys(serviceMapping)
    .map(service => writeImportStatement(service, true))
    .filter(Boolean)
    .join('\n')
}

function writeImportStatement(service, static = false) {
  const serviceOptions = options.services[service]
  if (serviceOptions.static !== static) {
    return
  }

  if (static) {
    return `import 'firebase/${serviceMapping[service]}'`
  }

  const webpackComments = []
  if (serviceOptions.chunkName) {
    webpackComments.push(`webpackChunkName: '${serviceOptions.chunkName}'`)
  }
  if (serviceOptions.preload) {
    webpackComments.push(`webpackPreload: true`)
  }
  let webpackCommentsString = ''
  if (webpackComments.length) {
    webpackCommentsString = `/* ${webpackComments.join(', ')} */`
  }
  return `await import(${webpackCommentsString}'firebase/${serviceMapping[service]}')`
}

<%= writeStaticImports() %>
...
    <%= writeImportStatement('auth') %>
    const fireAuth = firebase.auth()
    const fireAuthObj = firebase.auth
```

Regarding your questions:
- In production there is no benefit, but during development it could be helpful that webpack uses named chunk so that its immediately clear in your network console what each chunk represents. Its just a small DX improvement for debugging. In production mode these chunkNames should be ignored
- webpackPrefetch doesnt has much use in my opinion, you either want to lazy load or preload it, prefetching (downloading the js file but not executing it) doesnt sound useful. 
Webpack mode could be useful. Thinking about it, probably instead of the code above we could also just use `webpackMode: eager` to include to have the same effect as our `static: true/false` option above. But that should be tested to make sure it really works the same (webpackMode eager and a static import)

**Comment by lupas** [comment]
@pimlie 

Thanks for all the inputs and the feedback!
The suggestions have been implemented with [v3.2.0](https://github.com/lupas/nuxt-fire/releases/tag/v3.2.0) pretty much as you suggested them, thanks a lot.

I will have a look at adding the webpackMode option at another point in time, opened issue https://github.com/lupas/nuxt-fire/issues/67 for it.

**Comment by pimlie** [comment]
Thanks for your prompt work on this!
